### PR TITLE
Ac/starting malos for zwave

### PR DESCRIPTION
**Pull request by eljuguetero**
first candidate for Matrix Malos Zwave server

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+FIND_LIBRARY(READLINE_LIB NAMES readline)
+FIND_LIBRARY(ZMQ_LIB NAMES zmq)
+FIND_LIBRARY(MATRIX_LIB NAMES matrix_malos)
+FIND_LIBRARY(MATRIX_ZMQ_LIB NAMES matrix_malos_zmq)
+FIND_LIBRARY(ZWAVEIP_LIB NAMES zwaveip)
+FIND_LIBRARY(ZWCMDTOOL_LIB NAMES zw_cmd_tool)
+FIND_LIBRARY(ZWPARSE_LIB NAMES zw_parse_xml)
+FIND_LIBRARY(GFLAGS_LIB NAMES gflags)
+
+message(STATUS "ZMQ found => " "${ZMQ_LIB}")
```
Comment (arhuaco) [inline src/CMakeLists.txt:30]: We need a macro that integrates the "find_library" and the "message" call. It should fail if the library is not found.

File: src/driver_zwave.cpp [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+#include <chrono>
+#include <iostream>
+#include <string>
+#include <thread>
+
+#include <gflags/gflags.h>
+
+DEFINE_int32(port, 41230, "ZWaveIP gateway port");
+DEFINE_string(server, "::1", "ZWaveIP Gateway ");
+DEFINE_string(psk, "123456789012345678901234567890aa", "PSK");
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: What's the PSK?

File: src/driver_zwave.cpp [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+DEFINE_string(xml, "ZWave_custom_cmd_classes.xml", "XML zwave classes");
+
+#include "./driver_zwave.h"
+
+#include "./src/driver.pb.h"
+
+#define binaryCommand_BUFFER_SIZE 2000
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Weird constant name.

File: src/driver_zwave.cpp [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+
+void net_mgmt_command_handler(union evt_handler_struct evt) {
+  switch (evt.dsk_report.type) {
+    case APPROVE_REQUESTED_KEYS: {
+      matrix_malos::ZWaveDriver::requestedKeys_ =
+          evt.requested_keys.requested_keys;
+      matrix_malos::ZWaveDriver::csaInclusionRequested_ =
+          evt.requested_keys.csa_requested;
+
+      std::cout << "The joining node requests these keys:\n" << std::endl;
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Many references to std::cout. Shouldn't these be stderr?

File: src/driver_zwave.cpp [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+          std::cout << "and client side authentication" << std::endl;
+        }
+        std::cout << "Enter \'grantkeys\' to accept or \'abortkeys\' to cancel."
+                  << std::endl;
+      }
+      break;
+      case APPROVE_DSK: {
+        std::cout << "The joining node is reporting this device specific key:"
+                  << std::endl;
+        // print_hex_string(evt.dsk_report.dsk, 16);
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Delete comment? Or use a verbose flag to enable on demand.

File: src/driver_zwave.cpp [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+    case TRANSMIT_TIMEOUT:
+      std::cout << "\nTransmit attempt timed out\n";
+      break;
+  }
+}
+
+namespace matrix_malos {
+
+/* Static members of ZWaveDriver class */
+bool ZWaveDriver::panConnectionBusy_;
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Why using camelCase? In thise case it should be pan_connection_busy_

File: src/driver_zwave.cpp [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+uint8_t ZWaveDriver::requestedKeys_;
+uint8_t ZWaveDriver::csaInclusionRequested_;
+ZmqPusher* ZWaveDriver::static_zqm_push_update_;
+
+void ZresourceMDNSHelper() { zresource_mdns_thread_func(NULL); }
+
+ZWaveDriver::ZWaveDriver()
+    : MalosBase(kZWaveDriverName),
+      MDNSThread_(ZresourceMDNSHelper),
+      cfgPsk_(64) {
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Comment the constant 64 */ big foor to allow bar */

File: src/driver_zwave.cpp [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+  SetNeedsKeepalives(true);
+  SetMandatoryConfiguration(true);
+  SetNotesForHuman("ZWave Driver v1.0");
+  panConnectionBusy_ = false;
+
+  serverIP_ = FLAGS_server;
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: remember not to use camelCase :-) OTOH this flag doen't seem necessary. Is it?

File: src/driver_zwave.cpp [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+    Send(zwave);
+  } else if (zwave.operation() == ZwaveMsg::ADDNODE) {
+    AddNode();
+  } else if (zwave.operation() == ZwaveMsg::REMOVENODE) {
+    RemoveNode();
+  } else if (zwave.operation() == ZwaveMsg::SETDEFAULT) {
+    SetDefault();
+  } else if (zwave.operation() == ZwaveMsg::LIST) {
+    List();
+  }
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: else ... fatal error.

File: src/driver_zwave.cpp [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+bool ZWaveDriver::SendUpdate() { return true; }
+
+void ZWaveDriver::Send(ZwaveMsg& msg) {
+  std::string cmdName;
+  std::string className;
+
+  const zw_command_class* pClass;
+  const zw_command* pCmd;
+
+  cmdName = ZwaveCmdType_Name(msg.zwave_cmd().cmd());
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: define the string here.

const cmd_name = ....

sabe below. In general try to define the variable close to where it is used (C++ style).

File: src/driver_zwave.cpp [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+
+  pClass = zw_cmd_tool_get_class_by_name(className.c_str());
+  pCmd = zw_cmd_tool_get_cmd_by_name(pClass, cmdName.c_str());
+
+  if (!pClass || !pCmd) {
+    std::cerr << "Invalid <ZWave class, command> pair." << std::endl;
+    return;
+  }
+
+  static unsigned char binaryCommand[binaryCommand_BUFFER_SIZE];
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Mmm. Don't allocate in the stack?

unique_ptr<char> binary_command(new char ...);

binary_command.get() gives you access to the pointer.

File: src/driver_zwave.cpp [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+  memcpy(&binaryCommand[2], msg.zwave_cmd().params().c_str(),
+         msg.zwave_cmd().params().length());
+
+  int binaryCommandLen =
+      2 +
+      msg.zwave_cmd()
+          .params()
+          .length();  // sizeof([class_number,cmd_number,params])
+
+  if (0 != panConnectionBusy_) {
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: Inverted comparison? It's OK the way it is.

File: src/driver_zwave.cpp [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+    return;
+  }
+
+  if (msg.device() != destAddress_) {
+    if (panConnection_) {
+      zclient_stop(panConnection_);
+      panConnection_ = NULL;
+    }
+    // FIXME: Use thread synchronization instead of sleep to avoid "Socket
+    // Read
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: mangled comment?

File: src/driver_zwave.cpp [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+  if (zconnection_send_async(panConnection_, binaryCommand, binaryCommandLen,
+                             0)) {
+    panConnectionBusy_ = true;
+  }
+}
+
+void ZWaveDriver::AddNode() {
+  net_mgmt_learn_mode_start();
+
+  int idx = 0;
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: better to define idx below. before the first idx++

File: src/driver_zwave.cpp [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+  static uint8_t buf[200];
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: buf size should be more precise. It seems it's 6. I would use 8 maximum.

File: src/driver_zwave.cpp [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+  buf[idx++] = 0;
+  buf[idx++] = 0x07; /* ADD_NODE_S2 */
+  buf[idx++] = 0;    /* Normal power, no NWI */
+
+  zconnection_send_async(gwZipconnection_, buf, idx, 0);
+}
+
+void ZWaveDriver::RemoveNode() {
+  int idx = 0;
+  static uint8_t buf[200];
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: same comment about buffer size (can be much smaller). This happens in other places (see below).

File: src/driver_zwave.cpp [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+      if (ZwaveClassType_IsValid(n->info[i])) {
+        const std::string& className =
+            ZwaveClassType_Name(static_cast<ZwaveClassType>(n->info[i]));
+
+        ZwaveClassType zwave_class;
+        ZwaveClassType_Parse(className, &zwave_class);
+
+        node->add_zwave_class(zwave_class);
+
+        std::cout << "  " << std::hex << (int)n->info[i] << " " << className
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: static_cast<int>(...)

File: src/driver_zwave.cpp [hunk 18]
```diff
@@ -0,0 +1,452 @@
...
+
+  std::string buffer;
+  msg.SerializeToString(&buffer);
+  zqm_push_update_->Send(buffer);
+}
+
+bool ZWaveDriver::ConnectToGateway() {
+  gwZipconnection_ = ZipConnect(serverIP_.c_str());
+
+  if (gwZipconnection_) return true;
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: can gwZipconnection_ be null? Of so, compare to nullptr. Othersize just return gwZipconnection_;

File: src/driver_zwave.cpp [hunk 19]
```diff
@@ -0,0 +1,452 @@
...
+  } else {
+    std::cout << "ZWaveDriver connected to " << remote_addr << std::endl;
+  }
+  return zc;
+}
+
+void print_hex_string(const uint8_t* data, unsigned int datalen) {
+  unsigned int i;
+
+  for (i = 0; i < datalen; i++) {
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: define inside of the cycle.

File: src/driver_zwave.cpp [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+
+static int hex2int(char c) {
+  if (c >= '0' && c <= '9') {
+    return c - '0';
+  } else if (c >= 'a' && c <= 'f') {
+    return c - 'a' + 0xa;
+  } else if (c >= 'A' && c <= 'F') {
+    return c - 'A' + 0xa;
+  } else {
+    return -1;
```
Comment (arhuaco) [inline src/driver_zwave.cpp:414]: Isn't this a fatal error?

File: src/driver_zwave.cpp [hunk 21]
```diff
@@ -0,0 +1,452 @@
...
+  }
+}
+
+void ZWaveDriver::ParsePsk(const char* psk) {
+  int val;
```
Comment (arhuaco) [inline src/driver_zwave.cpp]: define inside of the loop

File: src/driver_zwave.cpp (hunk 20 above)
Comment (arhuaco) [reply src/driver_zwave.cpp:414]: It seems that this can happen when c is 0. Perhaps this should be handled outside?

File: src/driver_zwave.h [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+ public:
+  static bool panConnectionBusy_;
+  static uint8_t requestedKeys_;
+  static uint8_t csaInclusionRequested_;
+  // workaround for C callback compatibility
+  static ZmqPusher* static_zqm_push_update_;
+
+ private:
+  std::thread MDNSThread_;
+  zconnection* gwZipconnection_;
```
Comment (arhuaco) [inline src/driver_zwave.h]: camelCase

File: src/xml/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,18 @@
+add_subdirectory( test )
+include_directories( . )
+include_directories( "/usr/include/libxml2/" )
+
+include(FindPythonInterp)
+find_package(LibXml2)
+
+# Using gnu99 instead of c99 to get strdup(3)
```
Comment (arhuaco) [inline src/xml/CMakeLists.txt]: This is inherited from libxml2? (gnu99)

**Review by arhuaco — COMMENTED** [review]
Leaving comments.

**Review by arhuaco — APPROVED** [review]
Approving for the comments I left. Pending reviews...

**Review by yoelrc88 — APPROVED** [review]

**Review by kdpatino — APPROVED** [review]

**Pull request merged**